ctr/src/lib.rs

```diff
@@ -43,11 +43,193 @@
+};
+use core::fmt;
+use core::ops::Div;
+
+pub mod flavors;
+use flavors::CtrFlavor;
+
+/// Generic CTR block mode isntance.
+pub struct Ctr<B, F>
+where
+    B: BlockCipher,
+    B::BlockSize: Div<F::Size>,
+    Quot<B::BlockSize, F::Size>: ArrayLength<F>,
+    F: CtrFlavor,
+{
+    cipher: B,
+    nonce: GenericArray<F, Quot<B::BlockSize, F::Size>>,
+    counter: F,
+    buffer: Block<B>,
+    buf_pos: u8,
+}
+
+impl<B, F> Ctr<B, F>
+where
+    B: BlockCipher,
+    B::BlockSize: Div<F::Size>,
+    Quot<B::BlockSize, F::Size>: ArrayLength<F>,
+    F: CtrFlavor,
+{
+    fn check_data_len(&self, data: &[u8]) -> Result<(), LoopError> {
+        let bs = B::BlockSize::USIZE;
+        let leftover_bytes = bs - self.buf_pos as usize;
+        if data.len() < leftover_bytes {
+            return Ok(());
+        }
+        let blocks = 1 + (data.len() - leftover_bytes) / bs;
+        self.counter
+            .checked_add(blocks)
+            .ok_or(LoopError)
+            .map(|_| ())
+    }
+}
+
+impl<B, F> FromBlockCipher for Ctr<B, F>
+where
+    B: BlockCipher,
+    B::BlockSize: Div<F::Size>,
+    Quot<B::BlockSize, F::Size>: ArrayLength<F>,
+    F: CtrFlavor,
+{
+    type BlockCipher = B;
+    type NonceSize = B::BlockSize;
+
+    #[inline]
+    fn from_block_cipher(cipher: B, nonce: &Block<B>) -> Self {
+        let nonce = F::load(nonce);
+        Self {
+            cipher,
+            buffer: Default::default(),
+            nonce,
+            counter: Default::default(),
+            buf_pos: 0,
+        }
+    }
+}
+
+
+impl<B, F> SyncStreamCipher for Ctr<B, F>
+where
+    B: BlockCipher,
+    B::BlockSize: Div<F::Size>,
+    Quot<B::BlockSize, F::Size>: ArrayLength<F>,
+    F: CtrFlavor,
+{
+    fn try_apply_keystream(&mut self, mut data: &mut [u8]) -> Result<(), LoopError> {
+        self.check_data_len(data)?;
+        let bs = B::BlockSize::USIZE;
+        let pos = self.buf_pos as usize;
+        debug_assert!(bs > pos);
+
+        let mut counter = self.counter;
+        if pos != 0 {
+            if data.len() < bs - pos {
+                let n = pos + data.len();
+                xor(data, &self.buffer[pos..n]);
+                self.buf_pos = n as u8;
+                return Ok(());
+            } else {
+                let (l, r) = data.split_at_mut(bs - pos);
+                data = r;
+                xor(l, &self.buffer[pos..]);
+                counter.increment();
+            }
+        }
+        
+        // Process blocks in parallel if cipher supports it
+        let pb = B::ParBlocks::USIZE;
+        if pb != 1 {
+            let mut chunks = data.chunks_exact_mut(bs * pb);
+            let mut blocks: ParBlocks<B> = Default::default();
+            for chunk in &mut chunks {
+                for b in blocks.iter_mut() {
+                    *b = counter.generate_block(&self.nonce);
+                    counter.increment();
+                }
+
+                self.cipher.encrypt_blocks(&mut blocks);
```

Comment:
Try this:

```suggestion
                self.cipher.encrypt_par_blocks(&mut blocks);
```

The `encrypt_blocks` method was changed in cipher v0.3.0-pre to operate on a slice.